packages/@aws-cdk/aws-amplify/lib/app.ts

```diff
@@ -0,0 +1,287 @@
+   *
+   * @param scope
+   * @param id
+   * @param appId
+   *
+   * @returns Stubbed App
+   */
+  public static fromAppId(scope: Construct, id: string, appId: string): IApp {
+    class Import extends Resource implements IApp {
+      public readonly appId = appId;
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * App Id
+   *
+   * @attribute
+   */
+  public readonly appId: string;
+
+  /**
+   * App Name
+   *
+   * @attribute
+   */
+  public readonly appName: string;
+
+  /**
+   * App ARN
+   *
+   * @attribute
+   */
+  public readonly appArn: string;
+
+  /**
+   * App Default Domain
+   *
+   * @attribute
+   */
+  public readonly appDefaultDomain: string;
+
+  /**
+   * Resource
+   */
+  private readonly resource: CfnApp;
+
+  private readonly basicAuthResolver: BasicAuthResolver = new BasicAuthResolver();
+
+  private readonly environmentVariablesResolver: EnvironmentVariablesResolver = new EnvironmentVariablesResolver();
+
+  constructor(scope: Construct, id: string, props: AppProps) {
+    super(scope, id, props);
+
+    if (props.basicAuthConfig) {
+      this.setBasicAuth(props.basicAuthConfig.username, props.basicAuthConfig.password);
+    }
+
+    if (props.environmentVariables && props.environmentVariables.length > 0) {
+      this.environmentVariablesResolver.addEnvironmentVariables(...props.environmentVariables);
+    }
+
+    const resource = new CfnApp(scope, 'App', {
+      accessToken: props.accessToken,
+      basicAuthConfig: this.basicAuthResolver,
+      buildSpec: props.buildSpec,
+      customRules: props.customRules,
+      description: props.description,
+      environmentVariables: this.environmentVariablesResolver,
+      name: props.name,
+      oauthToken: props.oauthToken,
+      repository: props.repository,
+      tags: props.tags
+    });
+
+    if (props.iamServiceRole) {
+      resource.iamServiceRole = props.iamServiceRole.roleArn;
+    }
+
+    if (props.buildSpec && !props.iamServiceRole) {
+      // create lazy role
+      const role = new LazyRole(scope, 'lazy-role', {
+        assumedBy: new ServicePrincipal('amplify.amazonaws.com')
+      });
+
+      resource.iamServiceRole = role.roleArn;
+    }
+
+    this.appArn = resource.attrArn,
+    this.appDefaultDomain = resource.attrDefaultDomain;
+    this.appId = resource.attrAppId;
+    this.appName = resource.attrAppName;
+
+    this.resource = resource;
+  }
+
+  public validate(): string[] {
+    const errors: string[] = [];
+
+    if (this.resource.buildSpec && !this.resource.iamServiceRole) {
+      errors.push(`You need to specify a service role when using buildspecs`);
+    }
+
+    return errors;
+  }
+
+  /**
+   * Add Service Role
+   *
+   * @param role
+   */
+  public addServiceRole(role: IRole) {
+    this.resource.iamServiceRole = role.roleArn;
+  }
+
+  /**
+   * Set Basic Auth on App
+   *
+   * @param username
+   * @param password
+   */
+  public setBasicAuth(username: string, password: string) {
```

Comment:
Rename to “add”. We use “add” to represent that “you are adding this config to your resource definition”